src/type_parsers/datetime_parser.jl

```diff
@@ -49,108 +59,155 @@ function _unsafe_datetime(y=0, m=1, d=1, h=0, mi=0, s=0, ms=0)
+    if len - pos + 1 < 7
         (b != delim) && (code |= Parsers.INVALID)
         (pos >= len) && (code |= Parsers.EOF)
         return _unsafe_datetime(0), code, pos
     end
+    sign_mul = 1
+    if b == UInt8('-')
+        sign_mul = -1
+        pos += 1
+        b = buf[pos]
+    end
 
     year = 0
-    for i in 1:4
-        b -= 0x30
-        b > 0x09 && (return _unsafe_datetime(0), code | Parsers.INVALID, pos)
-        year = Int(b) + 10 * year
-        b = buf[pos += 1]
-        (i > 2 && b == UInt8('-')) && break
+    for i in 1:10
+        b0 = b - 0x30
+        b0 > 0x09 && (return _unsafe_datetime(0), code | Parsers.INVALID, pos)
+        year = Int(b0) + 10 * year
+        pos += 1
+        pos > len && (return _unsafe_datetime(0), code | Parsers.INVALID | Parsers.EOF, pos)
+        b = buf[pos]
+        b == UInt8('-') && break
     end
-    b != UInt8('-')  && (return _unsafe_datetime(year), code | Parsers.INVALID, pos)
-    b = buf[pos += 1]
+    year *= sign_mul
+    # If the year is larger than what can be represented by a 32 bit integer, fail to parse,
+    # values between typemin(Int32) and MIN_DATETIME are clamped to MIN_DATETIME
+    # values between typemax(Int32) and MAX_DATETIME are clamped to MAX_DATETIME
+    overflowed = (year > typemax(Int32) || year < typemin(Int32))
+    if b != UInt8('-') || overflowed
+        overflowed || (code |= Parsers.OVERFLOW)
+        return (_unsafe_datetime(year), code | Parsers.INVALID, pos)
+    end
+    pos += 1
+    pos > len && (return _unsafe_datetime(0), code | Parsers.INVALID | Parsers.EOF, pos)
+    b = buf[pos]
 
     month = 0
     for _ in 1:2
-        b -= 0x30
-        b > 0x09 && (return _unsafe_datetime(year), code | Parsers.INVALID, pos)
-        month = Int(b) + 10 * month
-        b = buf[pos += 1]
-        b == UInt8('-') && break
+        b0 = b - 0x30
+        b0 > 0x09 && (return _unsafe_datetime(year), code | Parsers.INVALID, pos)
+        month = Int(b0) + 10 * month
+        pos += 1
+        pos > len && (return _unsafe_datetime(year), code | Parsers.INVALID | Parsers.EOF, pos)
+        b = buf[pos]
     end
     month > 12 && (return _unsafe_datetime(year), code | Parsers.INVALID, pos)
-    b != UInt8('-')  && (return _unsafe_datetime(year, month), code | Parsers.INVALID, pos)
-    b = buf[pos += 1]
+    b != UInt8('-') && (return _unsafe_datetime(year, month), code | Parsers.INVALID, pos)
+    pos += 1
+    pos > len && (return _unsafe_datetime(year), code | Parsers.INVALID | Parsers.EOF, pos)
+    b = buf[pos]
 
     day = 0
-    for _ in 1:2
-        b -= 0x30
-        b > 0x09 && (return _unsafe_datetime(year, month), code | Parsers.INVALID, pos)
-        day = Int(b) + 10 * day
-        pos == len && (code |= Parsers.EOF; break)
-        b = buf[pos += 1]
-        (b == UInt8('T') ||  b == UInt8(' ')) && break
+    for i in 1:2
+        b0 = b - 0x30
+        b0 > 0x09 && (return _unsafe_datetime(year), code | Parsers.INVALID, pos)
```

Comment:
same question, why not this?
```suggestion
        b0 > 0x09 && (return _unsafe_datetime(year, month), code | Parsers.INVALID, pos)
```
(i know these are the invalid cases, so in some sense don't matter, but just want to be sure i'm following)

Reply:
Yeah I think I wanted only to provide values that are "valid", but got sloppy because we don't promise anything about invalid values... I'll go through these :+1: